### ISSUE TITLE
memory leakage for GPU training

### ISSUE THREAD

**Issue by Jiakui** [google/seq2seq#224]
I used the input files around 100 MB to train a model, memory usage will exceed 40 GB after one day running. Then the program would be killed. 

It seems that this problem is related to Tensorflow https://github.com/tensorflow/tensorflow/issues/9779 . 

Any one succeeded to overcome this problem?

Thanks.

**Comment by mohgh** [comment]
As described in tensorflow issue page, manual implementation of softmax will fix the problem. In `attention.py` replace
`scores_normalized = tf.nn.softmax(scores, name="scores_normalized")`
with
`scores_exp = tf.exp(scores)`
`scores_sum = tf.reduce_sum(tf.exp(scores), axis=0)`
 `scores_normalized = tf.truediv(scores_exp, scores_sum, name="scores_normalized")`

**Comment by ghost** [comment]
@mohgh 

Have you tried to implement softmax with your fix? I'm just asking because the fix on the [Tensorflow issue](https://github.com/tensorflow/tensorflow/issues/9779#issuecomment-300062628) seems different..

**Comment by mohgh** [comment]
Yes, I've fixed the issue this way and now memory usage is fixed on 1.3 gigabytes. The implemented code is the definition of softmax.

**Comment by ghost** [comment]
@mohgh 

Do you happen to know if this fix will influence BLEU scores? After implementing manual softmax I only get around 15 BLEU (whereas before, when my model managed to finish, I got 19 BLEU ?

**Comment by milanv1** [comment]
@mohgh 

Do you happen to know if this fix will influence BLEU scores? After implementing manual softmax I only get around 15 BLEU (whereas before, when my model managed to finish, I got 19 BLEU ?
